internal/jobcontainers/jobcontainer.go

```diff
@@ -0,0 +1,424 @@
+	if matched, err := regexp.MatchString(volumeGUIDRegex, s.Root.Path); !matched || err != nil {
+		return nil, fmt.Errorf(`invalid container spec - Root.Path '%s' must be a volume GUID path in the format '\\?\Volume{GUID}\'`, s.Root.Path)
+	}
+	if s.Root.Path[len(s.Root.Path)-1] != '\\' {
+		s.Root.Path += `\` // Be nice to clients and make sure well-formed for back-compat
+	}
+
+	container := newJobContainer(id, s)
+
+	// Create the job object all processes will run in.
+	options := &jobobject.Options{
+		Name:          fmt.Sprintf(jobContainerNameFmt, id),
+		Notifications: true,
+	}
+	job, err := jobobject.Create(ctx, options)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create job object")
+	}
+
+	// Parity with how we handle process isolated containers. We set the same flag which
+	// behaves the same way for a silo.
+	if err := job.SetTerminateOnLastHandleClose(); err != nil {
+		return nil, errors.Wrap(err, "failed to set terminate on last handle close on job container")
+	}
+	container.job = job
+
+	var path string
+	defer func() {
+		if err != nil {
+			container.Close()
+			if path != "" {
+				removeSandboxMountPoint(ctx, path)
+			}
+		}
+	}()
+
+	limits, err := specToLimits(ctx, id, s)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert OCI spec to job object limits")
+	}
+
+	// Set resource limits on the job object based off of oci spec.
+	if err := job.SetResourceLimits(limits); err != nil {
+		return nil, errors.Wrap(err, "failed to set resource limits")
+	}
+
+	// Setup directory sandbox volume will be mounted
+	sandboxPath := fmt.Sprintf(sandboxMountFormat, id)
+	if _, err := os.Stat(sandboxPath); os.IsNotExist(err) {
+		if err := os.MkdirAll(sandboxPath, 0777); err != nil {
+			return nil, errors.Wrap(err, "failed to create mounted folder")
+		}
+	}
+	path = sandboxPath
+
+	if err := mountSandboxVolume(ctx, path, s.Root.Path); err != nil {
+		return nil, errors.Wrap(err, "failed to bind payload directory on host")
+	}
+
+	container.sandboxMount = path
+	go container.waitBackground(ctx)
+	return container, nil
+}
+
+// CreateProcess creates a process on the host, starts it, adds it to the containers
+// job object and then waits for exit.
+func (c *JobContainer) CreateProcess(ctx context.Context, config interface{}) (_ cow.Process, err error) {
+	conf, ok := config.(*hcsschema.ProcessParameters)
+	if !ok {
+		return nil, errors.New("unsupported process config passed in")
+	}
+
+	absPath, commandLine, err := getApplicationName(conf.CommandLine, c.sandboxMount, os.Getenv("PATH"))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get application name from commandline %q", conf.CommandLine)
+	}
+
+	token, err := processToken(conf.User)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create process token")
+	}
+
+	env := append(os.Environ(), envMapToSlice(conf.Environment)...)
```

Comment:
Using `os.Environ` here will include any environment variables set in the shim, won't it? That seems undesirable. e.g. we don't want to pass the containerd TTRPC address to the container process.

Reply:
Oh yeah.. How would we wanna get around this, filter out some known env vars? That seems kind of fragile but not sure what else we could do.

Reply:
We should use `CreateEnvironmentBlock` to get a base environment block, convert to `[]string`, add in anything else that is specified in the container config, then use that.

Reply:
Side note: The comment on [`os/exec.Cmd.Env`](https://pkg.go.dev/os/exec#Cmd.Env) is incorrect for Windows. Go actually only uses `os.Environ()` when `Cmd.Env` is `nil` on non-Windows systems. On Windows they basically do what we want here where they call `CreateEnvironmentBlock` instead.

Perhaps the Go behavior for Windows should be changed, or at least the comment could be improved. It might also be nice if Go had some way on Windows (e.g. via `SysProcAttr`) to specify that we want the env block we pass be merged with the default block from `CreateEnvironmentBlock`.

Reply:
You can probably just copy the code from here for how to parse the environment block in Go: https://github.com/golang/go/blob/f21be2fdc6f1becdbed1592ea0b245cdeedc5ac8/src/internal/syscall/execenv/execenv_windows.go#L16-L54